### PR TITLE
[WIP] feat: add import dedicated server

### PR DESCRIPTION
**Pull request by majidkarimizadeh**

### CODE REVIEW

File: internal/core/ports/dedicated_server_repository.go [hunk 1]
```diff
@@ -26,4 +26,10 @@ type DedicatedServerRepository interface {
 		domain.ControlPanels,
 		*shared.RepositoryError,
 	)
+
+	// GetDedicatedServer return dedicated_server from the dedicated server api.
```
Comment (alrayyes) [inline internal/core/ports/dedicated_server_repository.go:30]: `returns data for a singular dedicated_server from the dedicated server api.` would be better here.

File: internal/core/ports/dedicated_server_service.go [hunk 2]
```diff
@@ -17,4 +17,7 @@ type DedicatedServerService interface {
 
 	// GetAllControlPanels gets dedicated_server.ControlPanels.
 	GetAllControlPanels(ctx context.Context) (domain.ControlPanels, *errors.ServiceError)
+
+	// GetDedicatedServer return dedicated server.
```
Comment (alrayyes) [inline internal/core/ports/dedicated_server_service.go:21]: See ports comment.

File: internal/core/services/dedicated_server/service_test.go [hunk 3]
```diff
@@ -183,3 +192,62 @@ func TestService_GetAllControlPanels(t *testing.T) {
 		},
 	)
 }
+
+func TestService_GetDedicatedServer(t *testing.T) {
```
Comment (alrayyes) [inline internal/core/services/dedicated_server/service_test.go:196]: `TestServiceGetDedicatedServer`

File: internal/core/services/dedicated_server/service_test.go [hunk 4]
```diff
@@ -183,3 +192,62 @@ func TestService_GetAllControlPanels(t *testing.T) {
...
+	t.Run(
+		"service passes back dedicated server from repository",
+		func(t *testing.T) {
+
```
Comment (alrayyes) [inline internal/core/services/dedicated_server/service_test.go:200]: For tests it's best to use the `arrange / act / assert` pattern and group that way. An example of this would be:

```go
id := "tralala"
want := domain.DedicatedServer{Id: id}
spy := repositorySpy{dedicatedServer: want}

got, err := service.GetDedicatedServer(context.TODO(), id)

assert.Nil(t, err)
assert.Equal(t, want, *got)
````

This way the developer knows exactly what belongs where in every test. TLDR: Lose the extra spaces :p

File: internal/core/services/dedicated_server/service_test.go [hunk 5]
```diff
@@ -183,3 +192,62 @@ func TestService_GetAllControlPanels(t *testing.T) {
...
+			)
+
+			_, err := service.GetDedicatedServer(context.TODO(), "id")
+
+			assert.Error(t, err)
+			assert.ErrorContains(t, err, "some error")
+		},
+	)
+
+	t.Run(
```
Comment (alrayyes) [inline internal/core/services/dedicated_server/service_test.go:235]: I don't immediately see the point of this test (help me out, I'm in a different timezone).

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
+package to_resource_model
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	domain "github.com/leaseweb/terraform-provider-leaseweb/internal/core/domain/dedicated_server"
+	resourcesModel "github.com/leaseweb/terraform-provider-leaseweb/internal/provider/resources/dedicated_server/model"
+)
+
+func AdaptDedicatedServer(dedicatedServer domain.DedicatedServer) resourcesModel.DedicatedServer {
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter.go:9]: Good idea to add a comment here as well as it's an exported function

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	domain "github.com/leaseweb/terraform-provider-leaseweb/internal/core/domain/dedicated_server"
+	resourceModel "github.com/leaseweb/terraform-provider-leaseweb/internal/provider/resources/dedicated_server/model"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_adaptRack(t *testing.T) {
+	rack := domain.NewRack("id", "cap", "type")
+
+	got := adaptRack(rack)
+	want := resourceModel.Rack{
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:16]: Nice

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+	pciCard := domain.NewPciCard("description")
+	got := adaptPciCard(pciCard)
+	want := resourceModel.PciCard{
+		Description: basetypes.NewStringValue("description"),
+	}
+	assert.Equal(t, want, got)
+}
+
+func Test_adaptPciCards(t *testing.T) {
+	pciCard1 := domain.NewPciCard("description1")
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:183]: It's better to use a mock here so this test doesn't have to be refactored every time `NewPciCard` changes. So I would make this something like `pciCards := {{description: "description1"}, {description: "description2"}}`

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	want := resourceModel.PciCards{
+		resourceModel.PciCard{Description: basetypes.NewStringValue("description1")},
+		resourceModel.PciCard{Description: basetypes.NewStringValue("description2")},
+	}
+	assert.Len(t, got, 2)
+	assert.Equal(t, want, got)
+}
+
+func Test_adaptHdds(t *testing.T) {
+	hdds := domain.Hdds{
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:196]: Shorthand can be used here:

```go
hdds := domain.Hdds{{Id: "id1"}, {Id: "id2"}}
```

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+	assert.Equal(t, want, got)
+}
+
+func Test_adaptSpecs(t *testing.T) {
+	specs := domain.NewSpecs(
+		"chassis",
+		true,
+		domain.Cpu{Quantity: 1},
+		domain.Ram{Size: 2},
+		domain.Hdds{domain.Hdd{Id: "id"}},
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:226]: shorthand can be used here as well.

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+func TestAdaptDedicatedServer(t *testing.T) {
+
+	specs := domain.NewSpecs(
+		"chassis",
+		true,
+		domain.Cpu{Quantity: 1},
+		domain.Ram{Size: 2},
+		domain.Hdds{domain.Hdd{Id: "id1"}},
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:245]: Shorthand :)

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+		domain.PciCards{domain.NewPciCard("d")},
+	)
+	featureAvailability := domain.FeatureAvailability{IpmiReboot: false}
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:248]: As these variables aren't used anymore it makes sense to just use the values themselves in the `New` function below, no?

File: internal/facades/dedicated_server/facade.go [hunk 13]
```diff
@@ -61,8 +66,25 @@ func (f DedicatedServerFacade) GetAllControlPanels(ctx context.Context) (
 	return &dataSourceControlPanels, nil
 }
 
+// GetDedicatedServer returns dedicated server details.
+func (f DedicatedServerFacade) GetDedicatedServer(ctx context.Context, id string) (
+	*resourceModel.DedicatedServer,
+	*shared.FacadeError,
+) {
+	dedicatedServer, err := f.dedicatedServerService.GetDedicatedServer(ctx, id)
```
Comment (alrayyes) [inline internal/facades/dedicated_server/facade.go:74]: You can call `GetDedicatedServer` in the `GetDedicatedServers` function now. Advantage is that you can refactor `TestService_GetDedicatedServers` to make the test simpler as you've already tested edge cases in `TestService_GetDedicatedServer`.

File: internal/provider/dedicated_server_resource_test.go [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package provider
+
+import (
+	"testing"
+)
+
+// TODO: Implement acceptance tests for import dedicated server.
+func TestAccDedicatedServerResource(t *testing.T) {
```
Comment (alrayyes) [inline internal/provider/dedicated_server_resource_test.go:8]: This is easy to do. See the public_cloud resource acceptance tests. It should be handled automagically. You can also check out the acceptance test part of the custom terraform provider tutorial.

File: internal/provider/resources/dedicated_server/dedicated_server/configure.go [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+	resp *resource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	coreClient, ok := req.ProviderData.(client.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/dedicated_server/configure.go:23]: Discussed this with @imtiazPabel. We should refactor this at some point as the code is repeated over and over again.

File: internal/provider/resources/dedicated_server/dedicated_server/resource.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/leaseweb/terraform-provider-leaseweb/internal/provider/client"
+)
+
+var (
+	_ resource.Resource                = &dedicatedServerResource{}
+	_ resource.ResourceWithConfigure   = &dedicatedServerResource{}
+	_ resource.ResourceWithImportState = &dedicatedServerResource{}
+	_ resource.ResourceWithModifyPlan  = &dedicatedServerResource{}
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/dedicated_server/resource.go:14]: Remove this interface implementation if you're not using it. The same for the other ones so we can get rid of the panics below.

File: internal/provider/resources/dedicated_server/dedicated_server/schema.go [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+			},
+			"null_routed": schema.BoolAttribute{
+				Computed: true,
+			},
+			"ports": portSchema(),
+		},
+	}
+}
+
+func portSchema() schema.ListNestedAttribute {
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/dedicated_server/schema.go:225]: nice

File: internal/provider/resources/dedicated_server/model/dedicated_server.go [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+import (
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type DedicatedServer struct {
+	Id                  types.String        `tfsdk:"id"`
+	AssetId             types.String        `tfsdk:"asset_id"`
+	SerialNumber        types.String        `tfsdk:"serial_number"`
+	Rack                Rack                `tfsdk:"rack"`
+	Location            Location            `tfsdk:"location"`
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/model/dedicated_server.go:12]: Are you sure this works?

**Review by alrayyes — COMMENTED** [review]
Looks good! I still highly recommend setting expected fields to `required` in the api definitions so that they are correct there.

File: internal/core/services/dedicated_server/service_test.go (hunk 3 above)
Comment (majidkarimizadeh) [reply internal/core/services/dedicated_server/service_test.go:196]: why!?
https://github.com/Leaseweb/terraform-provider-leaseweb/blob/4bafe06b199790e212aac2d0da21e1bc936d4a14/internal/core/services/public_cloud/service_test.go#L282

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go (hunk 10 above)
Comment (majidkarimizadeh) [reply internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:226]: making it long line, it is hard to read, I'd not do that in this case

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go (hunk 11 above)
Comment (majidkarimizadeh) [reply internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:245]: same, hard to read

File: internal/provider/dedicated_server_resource_test.go (hunk 14 above)
Comment (majidkarimizadeh) [reply internal/provider/dedicated_server_resource_test.go:8]: That's a good one @alrayyes, because I guess it's using the same variable in create to do the import and since we don't have create in dedicated server it's making it a bit tricky.

I checked the public cloud one:
https://github.com/Leaseweb/terraform-provider-leaseweb/blob/main/internal/provider/instance_resource_test.go#L86

there is no straightforward way to do the import with the dedicated server ID and checking the Terraform state.
pls correct me if I'm wrong.

File: internal/core/services/dedicated_server/service_test.go (hunk 5 above)
Comment (majidkarimizadeh) [reply internal/core/services/dedicated_server/service_test.go:235]: you're right <3

File: internal/provider/resources/dedicated_server/dedicated_server/resource.go (hunk 16 above)
Comment (majidkarimizadeh) [reply internal/provider/resources/dedicated_server/dedicated_server/resource.go:14]: I can get rid of only`ResourceWithModifyPlan`, for create/update/delete come with the `Resource` interface. pls correct me if I'm wrong.

File: internal/core/services/dedicated_server/service_test.go (hunk 4 above)
Comment (majidkarimizadeh) [reply internal/core/services/dedicated_server/service_test.go:200]: liked it <3

File: internal/facades/dedicated_server/facade.go (hunk 13 above)
Comment (majidkarimizadeh) [reply internal/facades/dedicated_server/facade.go:74]: in `GetDedicatedServers` we get the batch dedicated servers but in `GetDedicatedServer` we use ID to get one single resource, other than one function return `resourceModel` and the other one return `dataModel` so it seems merging these 2 making it more complicated! what do you think!?

File: internal/provider/resources/dedicated_server/model/dedicated_server.go (hunk 18 above)
Comment (majidkarimizadeh) [reply internal/provider/resources/dedicated_server/model/dedicated_server.go:12]: Basically we're still using the duplicated `dataModel` here in the resource directory, still we didn't add any custom login which make a differentiated between `resourceModel` and `dataModel`! also, we are looking into merging these 2 models if it's applicable! since we don't have tests for `dataModels` is it a good idea to add tests here? or wait for the result of merging investigation?!

File: internal/provider/resources/dedicated_server/dedicated_server/configure.go (hunk 15 above)
Comment (imtiazPabel) [reply internal/provider/resources/dedicated_server/dedicated_server/configure.go:23]: Basically, the Configure method is completely/almost identical regardless of every data source/ resources. We discussed that it would be better if we could write it once and use it everywhere.

File: internal/core/services/dedicated_server/service_test.go (hunk 3 above)
Comment (imtiazPabel) [reply internal/core/services/dedicated_server/service_test.go:185]: @alrayyes, through `IDE and go test` generator the method name comes same as @majidkarimizadeh defined

**Comment by imtiazPabel** [comment]
I'm missing the directory `examples/resources/leaseweb_dedicated_server/`

File: docs/resources/dedicated_servers.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "leaseweb_dedicated_servers Resource - leaseweb"
```
Comment (imtiazPabel) [inline docs/resources/dedicated_servers.md:3]: I feel, the resource should be singular, the existing `public_cloud_instnace` is also a singular. However, the legacy was also `leaseweb_dedicated_server`

File: docs/resources/dedicated_servers.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+- `port` (String)
+
+
+<a id="nestedatt--private_networks"></a>
+### Nested Schema for `private_networks`
+
+Read-Only:
+
+- `id` (String)
+- `link_speed` (Number)
```
Comment (imtiazPabel) [inline docs/resources/dedicated_servers.md:158]: Since link_speed is an enum, should be better to show the options.
https://github.com/Leaseweb/terraform-provider-leaseweb/blob/main/docs/resources/public_cloud_instance.md?plain=1#L74

An example: https://github.com/Leaseweb/terraform-provider-leaseweb/pull/50/files#diff-0db8933c108623e8a8857dccb481e5450918ed9a8395b6196a84213de80a58d2R169

File: docs/resources/dedicated_servers.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+
+<a id="nestedatt--rack"></a>
+### Nested Schema for `rack`
+
+Read-Only:
+
+- `capacity` (String)
+- `id` (String)
+- `type` (String)
```
Comment (imtiazPabel) [inline docs/resources/dedicated_servers.md:171]: https://github.com/Leaseweb/terraform-provider-leaseweb/pull/54/files#r1743360309

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	assert.Equal(t, want, got)
+}
+
+func Test_adaptPorts(t *testing.T) {
+	port1 := domain.NewPort("name1", "1111")
+	port2 := domain.NewPort("name2", "2222")
+	ports := domain.Ports{port1, port2}
+	got := adaptPorts(ports)
+	want := resourceModel.Ports{
```
Comment (imtiazPabel) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:79]: Since `resourceModel.Ports` is slice of `resourceModel.Port`. Should be enough

```
resourceModel.Ports{
		{Name: basetypes.NewStringValue("name1"), Port: basetypes.NewStringValue("1111")},
		{Name: basetypes.NewStringValue("name2"), Port: basetypes.NewStringValue("2222")},
	}
```

It's not a deal breaker. You can consider it as a comment 😊

File: internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go [hunk 23]
```diff
@@ -0,0 +1,275 @@
...
+		Status:    basetypes.NewStringValue("status"),
+		Subnet:    basetypes.NewStringValue("subnet"),
+		VlanId:    basetypes.NewStringValue("vlanid"),
+		LinkSpeed: basetypes.NewInt32Value(12),
+	}
+	assert.Equal(t, want, got)
+}
+
+func Test_adaptPrivateNetworks(t *testing.T) {
+	privateNetwork := domain.PrivateNetworks{
```
Comment (imtiazPabel) [inline internal/facades/dedicated_server/data_adapters/to_resource_model/adapter_test.go:111]: https://github.com/Leaseweb/terraform-provider-leaseweb/pull/54/files#r1743406380

File: internal/provider/resources/dedicated_server/dedicated_server/import_state.go [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+	resource.ImportStatePassthroughID(
+		ctx,
+		path.Root("id"),
+		req,
+		resp,
+	)
+
+	dedicatedServer, err := d.client.DedicatedServerFacade.GetDedicatedServer(ctx, req.ID)
+	if err != nil {
+		resp.Diagnostics.AddError("Error importing Dedicated server", err.Error())
```
Comment (imtiazPabel) [inline internal/provider/resources/dedicated_server/dedicated_server/import_state.go:25]: Adding the server ID in the message would be nice. So that, if there was, multiple schema for importing, customer can know for which server id import failed.

File: internal/provider/resources/dedicated_server/dedicated_server/schema.go [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+			},
+			"power_ports": portSchema(),
+			"private_networks": schema.ListNestedAttribute{
+				Computed: true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"id": schema.StringAttribute{
+							Computed: true,
+						},
+						"link_speed": schema.Int32Attribute{
```
Comment (imtiazPabel) [inline internal/provider/resources/dedicated_server/dedicated_server/schema.go:106]: Since In `api-defination`, the field is an `enum`. Would be good to validate.

File: internal/provider/resources/dedicated_server/model/doc.go [hunk 26]
```diff
@@ -0,0 +1,2 @@
+// Package model implements data models for dedicated_server data sources.
+package model
```
Comment (imtiazPabel) [inline internal/provider/resources/dedicated_server/model/doc.go:2]: Is it useful for any case?

**Comment by imtiazPabel** [comment]
If you are leaving more commits in the PR, I would say not to make `force` anymore. It makes tough to review, to see the specific changes. If you want to merge with a single commit, we can always `Squash & Merge`, in that case.

File: internal/core/services/dedicated_server/service_test.go (hunk 3 above)
Comment (alrayyes) [reply internal/core/services/dedicated_server/service_test.go:185]: Then it's good, nvm!

File: internal/provider/dedicated_server_resource_test.go (hunk 14 above)
Comment (alrayyes) [reply internal/provider/dedicated_server_resource_test.go:8]: I think the idea is that the import throws no errors in the tests. As this is all the tutorial does I'm assuming that it's recommended practice. I think you can [check the state](https://developer.hashicorp.com/terraform/plugin/testing/acceptance-tests), but you'll have to check the documentation. For now just running the import in acceptance tests should be fine.

File: internal/provider/resources/dedicated_server/dedicated_server/configure.go (hunk 15 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/dedicated_server/configure.go:23]: But that's for another PR, this is fine.

File: internal/provider/resources/dedicated_server/dedicated_server/resource.go (hunk 16 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/dedicated_server/resource.go]: That is correct.

File: internal/provider/resources/dedicated_server/model/dedicated_server.go (hunk 18 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/model/dedicated_server.go:12]: When you test this manually does it work? I don't think it should.

If above works then you don't need tests. Models are just lists of attributes with no extra logic. The only tested in resource models is `getAttributeTypes` (or whatever it's called). This method is used for object type metadata (see how public cloud models that implement other models are used).

File: docs/resources/dedicated_servers.md (hunk 19 above)
Comment (alrayyes) [reply docs/resources/dedicated_servers.md:3]: We discussed this. For end users this is trivial to change. I say let's do it properly and not keep old implementation just for the convenience of 3 people, while annoying all new users of terraform.

File: docs/resources/dedicated_servers.md (hunk 20 above)
Comment (alrayyes) [reply docs/resources/dedicated_servers.md:158]: Good catch @imtiazPabel! That is indeed the idea. Give the user as much information as possible beforehand. For stuff like this dynamic generation is even better (see examples).

File: internal/provider/resources/dedicated_server/dedicated_server/import_state.go (hunk 24 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/dedicated_server/import_state.go:25]: Let's write these rules down in the wiki somewhere.

File: internal/provider/resources/dedicated_server/model/doc.go (hunk 26 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/model/doc.go:2]: Put this in `doc.go` because the package has multiple files. This way the developer always  knows to edit `go.doc` if there are multiple files and package level documentation needs to be added.

File: docs/resources/dedicated_servers.md (hunk 19 above)
Comment (imtiazPabel) [reply docs/resources/dedicated_servers.md:3]: In terms of properly, I feel that, if we are dealing singular dedicated server then it should singular the resource name as well.

File: docs/resources/dedicated_servers.md (hunk 20 above)
Comment (majidkarimizadeh) [reply docs/resources/dedicated_servers.md:158]: This PR is about importing dedicated server! is this really matter in this PR!? there is no option to show to the customers, they can just import it!

File: internal/provider/resources/dedicated_server/dedicated_server/schema.go (hunk 25 above)
Comment (majidkarimizadeh) [reply internal/provider/resources/dedicated_server/dedicated_server/schema.go:106]: validate for import!? we don't get any input for this from user!

File: internal/provider/resources/dedicated_server/model/dedicated_server.go (hunk 18 above)
Comment (majidkarimizadeh) [reply internal/provider/resources/dedicated_server/model/dedicated_server.go:12]: that's what I get:

```sh
terraform import leaseweb_dedicated_server.imp 10836167

leaseweb_dedicated_server.imp: Importing from ID "10836167"...
leaseweb_dedicated_server.imp: Import prepared!
  Prepared leaseweb_dedicated_server for import
leaseweb_dedicated_server.imp: Refreshing state... [id=10836167]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```

---

```sh

terraform state show 'leaseweb_dedicated_server.imp'
# leaseweb_dedicated_server.imp:
resource "leaseweb_dedicated_server" "imp" {
    asset_id             = "10836167"
    contract             = {
...

```

**Comment by majidkarimizadeh** [comment]
> If you are leaving more commits in the PR, I would say not to make `force` anymore. It makes tough to review, to see the specific changes. If you want to merge with a single commit, we can always `Squash & Merge`, in that case.



> I'm missing the directory `examples/resources/leaseweb_dedicated_server/`

https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-documentation-generation#add-resource-import-documentation
done

**Comment by majidkarimizadeh** [comment]
> I'm missing the directory `examples/resources/leaseweb_dedicated_server/`

https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-documentation-generation#add-resource-import-documentation
done

File: .gitignore [hunk 27]
```diff
@@ -8,7 +8,7 @@ terraform.tfstate.backup
 *.log
 *.iml
 *.test
-
```
Comment (alrayyes) [inline .gitignore:11]: As this has nothing to do with the project it's the developers responsibility to put it in the [global .gitignore](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/) file. The project .gitignore should not care or know about which IDE the developer is using.

File: examples/resources/leaseweb_dedicated_server/resource.tf [hunk 28]
```diff
@@ -0,0 +1,4 @@
+# Manage example Dedicated server
+resource "leaseweb_dedicated_server" "example" {
+
```
Comment (alrayyes) [inline examples/resources/leaseweb_dedicated_server/resource.tf:3]: Are there really no required attributes for creating a dedicated_server?

File: internal/provider/dedicated_server_resource_test.go [hunk 29]
```diff
@@ -0,0 +1,24 @@
+package provider
+
+// TODO: fix the test case
```
Comment (alrayyes) [inline internal/provider/dedicated_server_resource_test.go:3]: tsk tsk tsk. Without tests this isn't done :p

File: internal/provider/resources/dedicated_server/dedicated_server/resource.go [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+	return &dedicatedServerResource{}
+}
+
+type dedicatedServerResource struct {
+	client client.Client
+}
+
+// Create implements resource.Resource.
+func (d *dedicatedServerResource) Create(context.Context, resource.CreateRequest, *resource.CreateResponse) {
+
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/dedicated_server/resource.go:26]: Create/Delete/Update do nothing. Is this intentional? If not see if you can maybe implement another interface that doesn't need these functions. If that's not possible and these should never be called then it's better to leave a panic statement in here imho.

File: internal/provider/resources/dedicated_server/dedicated_server/schema.go (hunk 25 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/dedicated_server/schema.go:106]: Fair point. We only need to validate if user can create/update/delete.

File: examples/resources/leaseweb_dedicated_server/resource.tf (hunk 28 above)
Comment (alrayyes) [reply examples/resources/leaseweb_dedicated_server/resource.tf:3]: I see now, there's is only for import. In that case a resource file isn't necessary as far as i know. The import.sh should be enough.

File: internal/provider/resources/dedicated_server/model/control_panels.go [hunk 31]
```diff
@@ -0,0 +1,5 @@
+package model
+
+type ControlPanels struct {
+	ControlPanels []ControlPanel `tfsdk:"control_panels"`
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/model/control_panels.go:4]: I think this will break once the user starts managing resources. But if it works for importing then it works!

File: internal/provider/resources/dedicated_server/model/dedicated_server.go (hunk 18 above)
Comment (alrayyes) [reply internal/provider/resources/dedicated_server/model/dedicated_server.go:12]: I'm pretty sure this will break once the user starts managing resources, but I could be wrong. But that's a problem for another day. I wouldn't merge data source & resource models just yet because of that.

File: internal/provider/resources/dedicated_server/model/specs.go [hunk 32]
```diff
@@ -0,0 +1,12 @@
+package model
+
+import "github.com/hashicorp/terraform-plugin-framework/types"
+
+type Specs struct {
+	Chassis             types.String `tfsdk:"chassis"`
+	HardwareRaidCapable types.Bool   `tfsdk:"hardware_raid_capable"`
+	Cpu                 Cpu          `tfsdk:"cpu"`
+	Ram                 Ram          `tfsdk:"ram"`
+	Hdds                Hdds         `tfsdk:"hdd"`
```
Comment (alrayyes) [inline internal/provider/resources/dedicated_server/model/specs.go:10]: `tfsdk:"hdds"`. Could you double check all the other stuff that should be plural?

**Review by alrayyes — CHANGES_REQUESTED** [review]
Terraform, the tool of a 1001 implementation details.

File: examples/resources/leaseweb_dedicated_server/resource.tf (hunk 28 above)
Comment (majidkarimizadeh) [reply examples/resources/leaseweb_dedicated_server/resource.tf:3]: True, `import.sh` is necessary one

**Comment by majidkarimizadeh** [comment]
> Create/Delete/Update do nothing. Is this intentional? If no...
> tsk tsk tsk. Without tests this isn't d....
> This is easy to do. See the public_cloud resource acceptance tests. It should...

@alrayyes since the messages on this PR is getting out of hand, let me clear the acceptance test issue.

There is no option to create the dedicated server via Leaseweb API. in this case, we can't implement create, and we just need to keep it empty. 
On the other side, we can't do much on the `import` acceptance test cases, if you check the Terraform documentations. it uses the same state on the create and tests import based on that.

BTW, I added a `TODO` to see if we can implement an acceptance test, but as I see it's not easy to implement in this situation, If you see an easy of implementing it LMK.

**Comment by majidkarimizadeh** [comment]
@alrayyes btw when I'm getting 67 comments on my PR while I was focused, it means something is wrong

**Comment by alrayyes** [comment]
Please, no more squashing commits while we're in the middle of a review. This is a lot of work now :p

File: internal/facades/dedicated_server/facade.go (hunk 13 above)
Comment (alrayyes) [reply internal/facades/dedicated_server/facade.go:74]: Ah, so there is a specific endpoint for batches? In that case then `GetDedicatedServer` should not be used.

**Review by alrayyes — CHANGES_REQUESTED** [review]
Looks good! Just a a couple of questions.

**Comment by majidkarimizadeh** [comment]
new implementation #61
